lib/vrl/DESIGN.md

```diff
@@ -0,0 +1,269 @@
+  "tricks" to making VRL fast that takeaway from readability.
+
+## Why VRL
+
+VRL exists to solve the need for a _safe_ and _performant_ **domain specific
+language** (DSL) to remap observability data.
+
+Its purpose is to hit a sweet spot between Turing complete languages such as
+Lua, and static transforms such as Vector's old `rename_fields`. It should be
+flexible enough to cover most remapping use-cases, without requiring the
+flexibility and downsides of a full programming language, such as poor
+readability and performance.
+
+See the [introduction blog post][blog] for more details on the **why**.
+
+[blog]: https://vector.dev/blog/vector-remap-language/#preamble
+
+## Target Audience
+
+VRL has a specialized purpose of solving observability data remapping needs.
+Because of this purpose, the language is mostly used by people who manage
+infrastructure at their organisations. The role of this group is usually
+referred to as "operator", "devops" (Development and Operations) or "SRE" (Site
+Reliability Engineer).
+
+One common generalization of this group is that they are focused on maintaining
+infrastructure within an organization, and often write and maintain their own
+software to achieve this goal. They are adept enough at programming to achieve
+their goals, but have no need or desire to be as skilled in programming as
+dedicated software engineers, because their time is best spent elsewhere.
+
+As with everything, there are exceptions to the rule, and many people in this
+group _are_ highly skilled software engineers, but VRL must capture the largest
+possible segment of this group, and should therefore be limited in complexity.
+
+Therefore, when extending the feature set of VRL, design **the feature for the
+intended target audience**, which will likely mean choosing different trade-offs
+than you'd make if you were to design the feature for your personal needs.
+
+## Language Limits
+
+There are a number of features that we've so far rejected to implement in the
+language. This might change in the future, but there should be a good reason to
+do so.
+
+- modules (see: [#5507][])
+
+  So far, we've had no need to split up functions over multiple modules. The
+  function naming rules make it so that most functions are already grouped by
+  their logical usage patterns.
+
+- classes
+
+  Given that VRL is a simple DSL, and that any indirection in a program's source
+  can lead to confusion, we've decided against introducing the concept of
+  classes, and instead focused on the usage of function calls to solve operator
+  needs.
+
+- user-defined functions
+
+  User-defined functions again produce indirection. While it _might_ be useful
+  to some extremely large use-cases, in most cases, allowing people to read
+  a program from top to bottom without having to jump around is more clear in
+  the context within which VRL is used (remapping).
+
+- network calls (see [#4517][] and [#8717][])
```

Comment:
Could this be expanded to general I/O, or do you think that differs?

Reply:
🤔 I think as we add enrichment functionality to VRL this will change. Imagining that we add an enrichment table for Redis and then call out to it from VRL, this could involve a network call. Granted VRL won't be doing the call, the enrichment table will, but it still affects throughput in the same way as VRL doing the network call itself.